File: src/protocol.ts
export class ProtocolError extends Error {}

export type ProjectorState = Record<string, unknown>;

type ProtobufField = {
  number: number;
  wireType: number;
  value: number | Buffer;
};

export function encodeVarint(value: number): Buffer {
  if (!Number.isSafeInteger(value) || value < 0) {
    throw new RangeError("varints must be non-negative safe integers");
  }
  const output: number[] = [];
  do {
    let byte = value & 0x7f;
    value = Math.floor(value / 128);
    if (value > 0) byte |= 0x80;
    output.push(byte);
  } while (value > 0);
  return Buffer.from(output);
}

export function decodeVarint(data: Buffer, start = 0): [number, number] {
  let value = 0;
  let multiplier = 1;
  let offset = start;
  while (offset < data.length && multiplier <= 2 ** 63) {
    const byte = data[offset++];
    if (byte === undefined) break;
    value += (byte & 0x7f) * multiplier;
    if ((byte & 0x80) === 0) {
      if (!Number.isSafeInteger(value)) throw new ProtocolError("varint exceeds safe integer range");
      return [value, offset];
    }
    multiplier *= 128;
  }
  throw new ProtocolError("invalid or incomplete varint");
}

export function fieldVarint(number: number, value: number): Buffer {
  return Buffer.concat([encodeVarint(number << 3), encodeVarint(value)]);
}

export function fieldBytes(number: number, value: Buffer): Buffer {
  return Buffer.concat([encodeVarint((number << 3) | 2), encodeVarint(value.length), value]);
}

export function frame(payload: Buffer): Buffer {
  return Buffer.concat([encodeVarint(payload.length), payload]);
}

export function keyPacket(keycode: number, pressed: boolean): Buffer {
  const event = Buffer.concat([fieldVarint(1, keycode), fieldVarint(2, Number(pressed))]);
  return frame(fieldBytes(2, fieldBytes(1, event)));
}

export function setVolumePacket(volume: number): Buffer {
  if (!Number.isInteger(volume) || volume < 0 || volume > 100) {
    throw new RangeError("volume must be an integer between 0 and 100");
  }
  const request = Buffer.from(JSON.stringify({ req: "setVolume", param: String(volume) }));
  const requestInfo = Buffer.concat([
    fieldBytes(1, Buffer.from("reqestinfo")),
    fieldBytes(2, request),
  ]);
  return frame(fieldBytes(2, fieldBytes(4, requestInfo)));
}

export function splitFrames(data: Buffer): { frames: Buffer[]; remainder: Buffer } {
  const frames: Buffer[] = [];
  let offset = 0;
  while (offset < data.length) {
    let length: number;
    let payloadStart: number;
    try {
      [length, payloadStart] = decodeVarint(data, offset);
    } catch (error) {
      if (error instanceof ProtocolError) break;
      throw error;
    }
    const payloadEnd = payloadStart + length;
    if (payloadEnd > data.length) break;
    frames.push(data.subarray(payloadStart, payloadEnd));
    offset = payloadEnd;
  }
  return { frames, remainder: data.subarray(offset) };
}

function parseFields(data: Buffer): ProtobufField[] {
  const fields: ProtobufField[] = [];
  let offset = 0;
  while (offset < data.length) {
    const [tag, afterTag] = decodeVarint(data, offset);
    offset = afterTag;
    const number = tag >> 3;
    const wireType = tag & 7;
    if (wireType === 0) {
      const [value, next] = decodeVarint(data, offset);
      fields.push({ number, wireType, value });
      offset = next;
    } else if (wireType === 2) {
      const [length, contentStart] = decodeVarint(data, offset);
      const end = contentStart + length;
      if (end > data.length) throw new ProtocolError("incomplete length-delimited field");
      fields.push({ number, wireType, value: data.subarray(contentStart, end) });
      offset = end;
    } else {
      throw new ProtocolError(`unsupported protobuf wire type ${wireType}`);
    }
  }
  return fields;
}

function bytesField(data: Buffer, number: number): Buffer {
  const field = parseFields(data).find(
    (candidate) => candidate.number === number && candidate.wireType === 2,
  );
  if (!field || !Buffer.isBuffer(field.value)) throw new ProtocolError(`missing field ${number}`);
  return field.value;
}

export function decodeStateFrame(payload: Buffer): [string, unknown] {
  const envelope = bytesField(payload, 3);
  const entry = bytesField(envelope, 1);
  const key = bytesField(entry, 1).toString("utf8");
  const rawValue = bytesField(entry, 2).toString("utf8");
  return [key, key === "sysconfig" ? JSON.parse(rawValue) : rawValue];
}

export function decodeState(data: Buffer): ProjectorState {
  const state: ProjectorState = {};
  for (const payload of splitFrames(data).frames) {
    try {
      const [key, value] = decodeStateFrame(payload);
      state[key] = value;
    } catch (error) {
      if (!(error instanceof ProtocolError) && !(error instanceof SyntaxError)) throw error;
    }
  }
  return state;
}

export function sanitizeState(value: unknown): unknown {
  if (typeof value === "string") return value.replace(/\p{Cf}/gu, "");
  if (Array.isArray(value)) return value.map(sanitizeState);
  if (value && typeof value === "object") {
    return Object.fromEntries(
      Object.entries(value).map(([key, item]) => [key, sanitizeState(item)]),
    );
  }
  return value;
}

export function redactState(state: ProjectorState): ProjectorState {
  const result = sanitizeState(state) as ProjectorState;
  const config = result.sysconfig;
  if (config && typeof config === "object" && !Array.isArray(config)) {
    result.sysconfig = { ...config, sn: "<redacted>", bluetooth_address: "<redacted>" };
  }
  return result;
}

File: src/jmgo_controller/protocol.py
from __future__ import annotations

import json
import unicodedata
from collections.abc import Iterator
from typing import Any


class ProtocolError(ValueError):
    """Raised when a JMGO packet cannot be decoded."""


def encode_varint(value: int) -> bytes:
    if value < 0:
        raise ValueError("varints must be non-negative")
    output = bytearray()
    while value > 0x7F:
        output.append((value & 0x7F) | 0x80)
        value >>= 7
    output.append(value)
    return bytes(output)


def decode_varint(data: bytes, offset: int = 0) -> tuple[int, int]:
    value = 0
    shift = 0
    while offset < len(data) and shift < 70:
        byte = data[offset]
        offset += 1
        value |= (byte & 0x7F) << shift
        if not byte & 0x80:
            return value, offset
        shift += 7
    raise ProtocolError("invalid or incomplete varint")


def field_varint(number: int, value: int) -> bytes:
    return encode_varint(number << 3) + encode_varint(value)


def field_bytes(number: int, value: bytes) -> bytes:
    return encode_varint((number << 3) | 2) + encode_varint(len(value)) + value


def frame(payload: bytes) -> bytes:
    return encode_varint(len(payload)) + payload


def key_packet(keycode: int, pressed: bool) -> bytes:
    event = field_varint(1, keycode) + field_varint(2, int(pressed))
    wrapped_event = field_bytes(1, event)
    return frame(field_bytes(2, wrapped_event))


def set_volume_packet(volume: int) -> bytes:
    if not 0 <= volume <= 100:
        raise ValueError("volume must be between 0 and 100")
    request = json.dumps({"req": "setVolume", "param": str(volume)}, separators=(",", ":")).encode()
    request_info = field_bytes(1, b"reqestinfo") + field_bytes(2, request)
    return frame(field_bytes(2, field_bytes(4, request_info)))


def iter_frames(data: bytes) -> Iterator[bytes]:
    offset = 0
    while offset < len(data):
        length, payload_start = decode_varint(data, offset)
        payload_end = payload_start + length
        if payload_end > len(data):
            raise ProtocolError("incomplete frame")
        yield data[payload_start:payload_end]
        offset = payload_end


def parse_fields(data: bytes) -> list[tuple[int, int, int | bytes]]:
    fields: list[tuple[int, int, int | bytes]] = []
    offset = 0
    while offset < len(data):
        tag, offset = decode_varint(data, offset)
        number, wire_type = tag >> 3, tag & 7
        if wire_type == 0:
            value, offset = decode_varint(data, offset)
        elif wire_type == 2:
            length, offset = decode_varint(data, offset)
            end = offset + length
            if end > len(data):
                raise ProtocolError("incomplete length-delimited field")
            value = data[offset:end]
            offset = end
        else:
            raise ProtocolError(f"unsupported protobuf wire type {wire_type}")
        fields.append((number, wire_type, value))
    return fields


def _only_bytes_field(data: bytes, number: int) -> bytes:
    for field_number, wire_type, value in parse_fields(data):
        if field_number == number and wire_type == 2 and isinstance(value, bytes):
            return value
    raise ProtocolError(f"missing field {number}")


def decode_state_frame(payload: bytes) -> tuple[str, Any]:
    envelope = _only_bytes_field(payload, 3)
    entry_wrapper = _only_bytes_field(envelope, 1)
    key = _only_bytes_field(entry_wrapper, 1).decode("utf-8")
    raw_value = _only_bytes_field(entry_wrapper, 2).decode("utf-8")
    if key == "sysconfig":
        return key, json.loads(raw_value)
    return key, raw_value


def decode_state(data: bytes) -> dict[str, Any]:
    state: dict[str, Any] = {}
    for payload in iter_frames(data):
        try:
            key, value = decode_state_frame(payload)
        except (ProtocolError, UnicodeDecodeError, json.JSONDecodeError):
            continue
        state[key] = value
    return state


def sanitize_state(value: Any) -> Any:
    if isinstance(value, str):
        return "".join(character for character in value if unicodedata.category(character) != "Cf")
    if isinstance(value, dict):
        return {key: sanitize_state(item) for key, item in value.items()}
    if isinstance(value, list):
        return [sanitize_state(item) for item in value]
    return value


def redact_state(state: dict[str, Any]) -> dict[str, Any]:
    result = sanitize_state(state)
    config = result.get("sysconfig")
    if isinstance(config, dict):
        config = dict(config)
        for key in ("sn", "bluetooth_address"):
            if key in config:
                config[key] = "<redacted>"
        result["sysconfig"] = config
    return result
